### PR TITLE
Issue #1262 - enable wildcard search for _include, _revinclude

### PR DESCRIPTION
**Pull request by michaelwschroeder**
Signed-off-by: Mike Schroeder <mschroed@us.ibm.com>

### CODE REVIEW

File: fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java [hunk 1]
```diff
@@ -20,9 +20,12 @@
 
 import org.testng.annotations.Test;
 
+import com.ibm.fhir.model.resource.Condition;
```
Comment (prb112) [inline fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java:23]: Update Copyright please

File: fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java [hunk 2]
```diff
@@ -327,4 +340,151 @@ public void testMultiIncludeRevinclude() throws Exception {
...
+        }
+
+        assertFalse(searchContext.hasIncludeParameters());
+
+        String selfUri = SearchUtil.buildSearchSelfUri("http://example.com/Patient", searchContext);
+        assertTrue(selfUri.contains(include1));
+        assertTrue(selfUri.contains(include2));
+        assertTrue(selfUri.contains(include3));
+    }
+
```
Comment (prb112) [inline fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java:489]: Having an integration test in SearchTest.java in fhir-server-test would be ideal

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 3]
```diff
@@ -502,6 +502,47 @@ private static SearchParameter getSearchParameterByUrlIfPresent(Map<String, Para
...
+     * @return
+     *         the inclusion SearchParameters for type {@code resourceType} or empty map if none exist
+     * @throws Exception
+     */
+    private static Map<String, SearchParameter> getInclusionWildcardSearchParameters(String resourceType, String joinResourceType,
+        String searchParameterTargetType, String inclusionKeyword) throws Exception {
+        Map<String, SearchParameter> inclusionSearchParameters = new HashMap<>();
+
+        for (SearchParameter searchParameter : getApplicableSearchParameters(joinResourceType)) {
+            if (searchParameter.getType().getValue().equals("reference") &&
```
Comment (tbieste) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:532]: Looks like SearchParamType.REFERENCE.equals(...) could be done instead against "reference".  Same comment on line 1485.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 4]
```diff
@@ -502,6 +502,47 @@ private static SearchParameter getSearchParameterByUrlIfPresent(Map<String, Para
...
+                    ((SearchConstants.INCLUDE.equals(inclusionKeyword)
+                            && (searchParameterTargetType == null || isValidTargetType(searchParameterTargetType, searchParameter))) ||
+                    (SearchConstants.REVINCLUDE.equals(inclusionKeyword) && isValidTargetType(resourceType, searchParameter)))) {
+                // Valid search parameter of type reference - add to map
+                inclusionSearchParameters.put(searchParameter.getCode().getValue(), searchParameter);
+            } else if (inclusionSearchParameters.containsKey(searchParameter.getCode().getValue())) {
+                // Search parameter is not valid - remove if search parameter by same name is alread in map
```
Comment (tbieste) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:539]: Typo in word "already".

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 5]
```diff
@@ -502,6 +502,47 @@ private static SearchParameter getSearchParameterByUrlIfPresent(Map<String, Para
...
+                inclusionSearchParameters.remove(searchParameter.getCode().getValue());
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:540]: Why would there be two search parameters of the same code already in the map? 

The word already is misspelled

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 6]
```diff
@@ -1373,17 +1414,25 @@ public static String normalizeForSearch(String value) {
     }
 
     /**
-     * Parses _include and _revinclude search result parameters contained in the
-     * query string, and produces
-     * InclusionParameter objects to represent those parameters. The
-     * InclusionParameter objects are included in the
-     * appropriate collections encapsulated in the passed FHIRSearchContext.
+     * Parses _include and _revinclude search result parameters contained in the query string, and produces
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1417]:
```suggestion
     * Parses _include and _revinclude search parameters contained in the query string, and produces
```

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 7]
```diff
@@ -1399,66 +1448,73 @@ private static void parseInclusionParameter(Class<?> resourceType, FHIRSearchCon
...
             } else {
-                newInclusionParm =
-                        buildRevIncludeParameter(resourceType, joinResourceType, searchParm, searchParameterName,
-                                searchParameterTargetType);
-                context.getRevIncludeParameters().add(newInclusionParm);
+                searchParm = getSearchParameter(joinResourceType, searchParameterName);
+                if (searchParm == null) {
+                    String msg = "Undefined Inclusion Parameter: " + inclusionValue;
+                    if (lenient) {
+                        // TODO add this to the list of supplemental warnings?
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1478]: todo is necessary?

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 8]
```diff
@@ -1399,66 +1448,73 @@ private static void parseInclusionParameter(Class<?> resourceType, FHIRSearchCon
...
 
-            if (inclusionKeyword.equals(SearchConstants.INCLUDE)) {
-                newInclusionParms =
-                        buildIncludeParameter(resourceType, joinResourceType, searchParm, searchParameterName,
-                                searchParameterTargetType);
-                context.getIncludeParameters().addAll(newInclusionParms);
+            // Ensure that the Inclusion Parameter being parsed is a valid search parameter of type 'reference'.
+            Map<String, SearchParameter> searchParametersMap;
+            if (SearchConstants.WILDCARD.equals(searchParameterName)) {
+                searchParametersMap = getInclusionWildcardSearchParameters(resourceType.getSimpleName(), joinResourceType, searchParameterTargetType, inclusionKeyword);
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1472]: What happens when it's set to lenient? e.g. do we fail it? look at the behavior on line 1477

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 9]
```diff
@@ -1399,66 +1448,73 @@ private static void parseInclusionParameter(Class<?> resourceType, FHIRSearchCon
...
+                        // TODO add this to the list of supplemental warnings?
+                        log.fine(msg);
+                        continue;
+                    } else {
+                        throw SearchExceptionUtil.buildNewInvalidSearchException(msg);
+                    }
+                }
+                if (!searchParm.getType().getValue().equals("reference")) {
+                    throw SearchExceptionUtil.buildNewInvalidSearchException("Inclusion Parameter must be of type 'reference'. "
+                            + "The passed Inclusion Parameter is of type: " + searchParm.getType().getValue());
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1487]: single quotes around the value.  it shows better in the logs

File: fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java [hunk 10]
```diff
@@ -327,4 +340,151 @@ public void testMultiIncludeRevinclude() throws Exception {
...
+        assertEquals("context", incParm.getSearchParameter());
+        assertEquals("Encounter", incParm.getSearchParameterTargetType());
+        assertFalse(searchContext.hasIncludeParameters());
+
+        String selfUri = SearchUtil.buildSearchSelfUri("http://example.com/Encounter", searchContext);
+        assertTrue(selfUri.equals("http://example.com/Encounter?_count=10&_revinclude=" + incParm.getJoinResourceType() +
+            ":" + incParm.getSearchParameter() + ":" + incParm.getSearchParameterTargetType() + "&_page=1"));
+    }
+
+    @Test
```
Comment (prb112) [inline fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java:459]: Let's add a set of extra tests. That has two _include search parameters on Patient and Provenance.  we want to double check that it works across multiple resources.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 8 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1472]: In the case of a wildcard search, if no valid matching search parameters are found, then we simply don't generate an _include or _revinclude clause - no exception is thrown. Should an exception be thrown in that case (if non-lenient)? I can certainly add a log entry indicating no valid search parms were found.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 5 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:540]: This is covering a case where there's a built-in search parm of type 'reference', but there's a tenant-specific search parm of the same name and the type is not 'reference'. I thought the tenant-specific parm should win.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 8 above)
Comment (prb112) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1472]: Maybe we just log out, if it is set to not lenient?

Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1472]: Yeah, will just log it.

File: fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java (hunk 1 above)
Comment (michaelwschroeder) [reply fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java:23]: Fixed.

File: fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java (hunk 2 above)
Comment (michaelwschroeder) [reply fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java:523]: Test added.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 3 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java]: Fixed.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 4 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java]: Fixed.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 5 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java]: Added log message and did not remove existing map entry.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 6 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java]: Fixed.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 9 above)
Comment (michaelwschroeder) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java]: Fixed.

File: fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java (hunk 10 above)
Comment (michaelwschroeder) [reply fhir-search/src/test/java/com/ibm/fhir/search/test/InclusionParameterParseTest.java:493]: Added new unit tests.

**Comment by prb112** [comment]
Fix https://github.com/IBM/FHIR/pull/1492

**Review by tbieste — APPROVED** [review]

**Review by prb112 — APPROVED** [review]
LGTM

**Pull request merged**